politeiawww/dcc.go

```diff
@@ -538,3 +574,273 @@ func (p *politeiawww) processGetDCCs(gds cms.GetDCCs) (*cms.GetDCCsReply, error)
 		DCCs: dccs,
 	}, nil
 }
+
+func (p *politeiawww) processSupportOpposeDCC(sd cms.SupportOpposeDCC, u *user.User) (*cms.SupportOpposeDCCReply, error) {
+	log.Tracef("processSupportOpposeDCC: %v %v", sd.Token, u.ID)
+
+	dcc, err := p.getDCC(sd.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode: cms.ErrorStatusInvoiceNotFound,
+			}
+		}
+		return nil, err
+	}
+
```

Comment:
These user errors `ErrorStatusUserActionNotAllowed` don't tell the user what they did wrong.

Reply:
How detailed would you like be to be for these user action errors?  Should be an extra 3-4 errors instead of using the catch all UserActionNotAllowed

Reply:
OK added ErrorStatusInvalidSupportOppose, ErrorStatusUserDuplicateSupportOppose and
ErrorStatusUserCannotSupportOpposeOwn instead of simply UserActionNotAllowed.

Reply:
So I try to make user errors somewhat generic so that you don't end up will a million different user errors and then use the `ErrorContext` field to fill in the details.  If I was writing this I would do something like:

```
	// The submitted Vote in the request must either be "aye" or "nay"
	if sd.Vote != supportString && sd.Vote != opposeString {
		return nil, www.UserError{
			ErrorCode:    cms.ErrorStatusInvalidSupportOppose,
			ErrorContext: []string{"support string not aye or nay"},
		}
	}

	// Check to make sure the user has not SupportOpposeed or Opposed this DCC yet
	if stringInSlice(dcc.SupportUserIDs, u.ID.String()) ||
		stringInSlice(dcc.OppositionUserIDs, u.ID.String()) {
		return nil, www.UserError{
			ErrorCode: cms.ErrorStatusDuplicateSupportOppose,
		}
	}

	// Check to make sure the user is not the author of the DCC.
	if dcc.SponsorUserID == u.ID.String() {
		return nil, www.UserError{
			ErrorCode: cms.ErrorStatusUserIsAuthor,
		}
	}

	// Check to make sure that the DCC is still active
	if dcc.Status != cms.DCCStatusActive {
		return nil, www.UserError{
			ErrorCode:    cms.ErrorStatusWrongDCCStatus,
			ErrorContext: []string{"dcc status must be active"},
		}
	}
```

Error statuses like `UserIsAuthor` and `WrongDCCStatus` are generic enough that they can be reused in other routes as well with the `ErrorContext` containing the specifics.

Reply:
Resolved as you suggested above.  Thanks.